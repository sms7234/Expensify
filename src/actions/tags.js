import database from '../firebase/firebase';

export const setTags = (tags) => ({
  type: "SET_TAGS",
  tags
});

export const startSetTags = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/tags`)
      .once('value')
      .then((snapshot) => {
        const tags = [];
        snapshot.forEach((child) => {
          tags.push({
            id: child.key,
            ...child.val()
          });
        });
        dispatch(setTags(tags));
      });
  };
};

export const addTag = (tag) =>({
  type: "ADD_TAG",
  tag
});

export const startAddTag = (tagData={}) => {
  return (dispatch, getState) => {
    const {
      tag='',
      description=''
    } = tagData;
    const tagItem={tag,description};
    const uid=getState().auth.uid;
    return database.ref(`users/${uid}/tags`).push(tagItem).then((ref)=>{
      dispatch(addTag({
        id: ref.key,
        ...tagItem
      }));
    });
  };
};

export const editTag = (id, updates) => ({
  type: 'EDIT_TAG',
  id,
  updates
});

export const startEditTag = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/tags/${id}`).update(updates).then(()=>{
      dispatch(editTag(id,updates));
    });
  };
};

export const removeTag = (id) => ({
  type: 'REMOVE_TAG',
  id
});

export const startRemoveTag = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/tags/${id}`).remove().then(()=>{
      dispatch(removeTag(id));
    });
  };
};
