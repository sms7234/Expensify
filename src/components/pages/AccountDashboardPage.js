import React from 'react';
import AccountList from '../lists/AccountList';
import AccountSummary from '../independents/AccountSummary';
import AccountListFilters from '../filters/AccountListFilters';

const AccountDashboardPage = () => (
  <div>
    <AccountSummary />
    <AccountListFilters />
    <AccountList />
  </div>
)

export default AccountDashboardPage;
