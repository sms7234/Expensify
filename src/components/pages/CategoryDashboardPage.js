import React from 'react';
import CategoryList from '../lists/CategoryList';
import CategoryListFilters from '../filters/CategoryListFilters';
import CategorySummary from '../independents/CategorySummary';

const CategoryDashboardPage = () => (
  <div>
    <CategorySummary />
    <CategoryListFilters />
    <CategoryList />
  </div>
)

export default CategoryDashboardPage;
