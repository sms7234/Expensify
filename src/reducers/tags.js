export default (state=[], action) => {
  switch(action.type){
    case 'SET_TAGS':
      return action.tags;
    case 'ADD_TAG':
      return state.concat(action.tag);
    case 'EDIT_TAG':
      return state.map((item) => {
        if (item.id === action.id){
          return {
            ...item,
            ...action.updates
          }
        } else {return item}
      });
      case 'REMOVE_TAG':
        return state.filter(({id}) => id!==action.id);
    default:
      return state;
  }
}
