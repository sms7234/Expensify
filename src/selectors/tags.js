export default (tags, {tag}) => {
  return tags.filter((item) => {
    const textMatch = item.tag.toLowerCase().includes(tag.toLowerCase());
    return textMatch;
  }).sort((a,b)=> {
    if(a.tag < b.tag){
      return -1;
    } else{
      return 1;
    }
  });
};
