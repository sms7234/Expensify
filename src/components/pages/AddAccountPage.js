import React from 'react';
import AccountForm from '../forms/AccountForm';
import {connect} from 'react-redux';
import {startAddAccount} from '../../actions/accounts';

export class AddAccountPage extends React.Component{
  onSubmit=(account) => {
    this.props.startAddAccount(account);
    this.props.history.push('/accounts');
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title"> Add Account </h1>
          </div>
        </div>
        <div className="content-container">
          <AccountForm onSubmit={this.onSubmit} />
        </div>
      </div>
    );
  };
};

const mapDispatchToProps = (dispatch) => ({
  startAddAccount:(account) => dispatch(startAddAccount(account))
});

export default connect(undefined, mapDispatchToProps)(AddAccountPage);
