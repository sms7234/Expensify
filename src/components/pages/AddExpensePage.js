import React from 'react';
import ExpenseForm from '../forms/ExpenseForm';
import { connect} from 'react-redux';
import {startAddExpense} from '../../actions/expenses';

export class AddExpensePage extends React.Component{
  onSubmit = (expense) => {
    this.props.startAddExpense(expense);
    this.props.history.push('/expenses');
  };
  render(){
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title"> Add Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm
            onSubmit = {this.onSubmit}
            categoryList = {this.props.categoryList}
            accountList = {this.props.accountList}
            tagList = {this.props.tagList}
          />
        </div>
      </div>
    );
  }
};

const mapStateToProps=(state) => {
  return {
    accountList: state.accounts.map((item)=>{return({label: item.account, value: item.account})}),
    categoryList: state.categories.map((item)=>{return({label: item.category, value: item.category})}),
    tagList: state.tags.map((item)=>{return({label: item.tag, value: item.tag})})
  };
};

const mapDispatchToProps = (dispatch) => ({
  startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(mapStateToProps, mapDispatchToProps)(AddExpensePage);
