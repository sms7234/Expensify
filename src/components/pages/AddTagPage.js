import React from 'react';
import TagForm from '../forms/TagForm';
import {connect} from 'react-redux';
import {startAddTag} from '../../actions/tags';

export class AddTagPage extends React.Component{
  onSubmit=(tag) => {
    this.props.startAddTag(tag);
    this.props.history.push('/tags');
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title"> Add Tag </h1>
          </div>
        </div>
        <div className="content-container">
          <TagForm onSubmit={this.onSubmit} />
        </div>
      </div>
    );
  };
};

const mapDispatchToProps = (dispatch) => ({
  startAddTag:(tag) => dispatch(startAddTag(tag))
});

export default connect(undefined, mapDispatchToProps)(AddTagPage);
