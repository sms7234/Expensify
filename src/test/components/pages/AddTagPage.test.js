import React from 'react';
import {shallow} from 'enzyme';
import {AddTagPage} from '../../../components/pages/AddTagPage';
import tags from '../../fixtures/tags';

let startAddTag, history, wrapper;

beforeEach(()=> {
  startAddTag = jest.fn();
  history = {push: jest.fn()};
  wrapper = shallow(<AddTagPage startAddTag={startAddTag} history={history} />);
});

test('should render AddTagPage correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle onSubmit',() => {
  wrapper.find('TagForm').prop('onSubmit')(tags[0]);
  expect(history.push).toHaveBeenLastCalledWith('/tags');
  expect(startAddTag).toHaveBeenLastCalledWith(tags[0]);
})
