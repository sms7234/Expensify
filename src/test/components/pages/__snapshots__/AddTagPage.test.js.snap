// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddTagPage correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
         Add Tag 
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <TagForm
      onSubmit={[Function]}
    />
  </div>
</div>
`;
